### ISSUE TITLE
On-device-training fails using Tflite_runtime: Node number 54 (FlexReluGrad) failed to prepare.

### ISSUE THREAD

**Issue by ahlzouao** [tensorflow/tensorflow#54244]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **Yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Custom embedded Linux distribution (Kernel 5.15)**
- Mobile device: **Raspberry Pi 4B**
- TensorFlow installed from (source or binary): **Source (Building on Tflite_runtime using pip_package scripts)**
- TensorFlow version (use command below): **2.7.0**
- Python version: **3.8**
- Bazel version (if compiling from source): **N/A**
- GCC/Compiler version (if compiling from source): **9.3 (OpenEmbedded GNU Toolchain)**
- CUDA/cuDNN version: **N/A**
- GPU model and memory: **N/A**

**Describe the current behavior**
I am following the tutorial on how to do [on-device-training](https://www.tensorflow.org/lite/examples/on_device_training/overview). The first step was to create and train the Fashion_mnist model on google Colab which was successful since I managed to download as an output the tflite model after converting (I made sure to mention all the necessary signatures while saving the model). The documentation is providing an example only on Java for android Apps, but I'm trying to explore whether this is feasible with tflite_runtime wheel.
I sent then the tflite model to the target (raspberry pi) where the Tflite_runtime has been installed using [this article](https://www.tensorflow.org/lite/guide/build_cmake_pip) and I'm trying to run the training function by feeding my neural network with arrays of zeros just to prove that it is working. This is the code snippet I'm running on my target. 

```
import numpy as np
import struct as st
import tflite_runtime.interpreter as tflr
nImg = 10000
nR = 28
nC =28
images_array = np.zeros((1, 28, 28), dtype="float32") 
labels_array = np.zeros((1,10), dtype="float32")
modelpath="fashion_mnist_model.tflite"
interpreter = tflr.Interpreter(modelpath)
signatures = interpreter.get_signature_list()
```

```
print('Signature:', signatures)
>>> Signature: {'infer': {'inputs': ['x'], 'outputs': ['logits', 'output']}, 'restore': {'inputs': ['checkpoint_path'], 'outputs': ['dense_1/bias:0', 'dense_1/kernel:0', 'dense_2/bias:0', 'dense_2/kernel:0']}, 'save': {'inputs': ['checkpoint_path'], 'outputs': ['checkpoint_path']}, 'train': {'inputs': ['x', 'y'], 'outputs': ['loss']}}
```
```
train = interpreter.get_signature_runner('train')
infer = interpreter.get_signature_runner('infer')
train(x=images_array, y=labels_array)
```
The script fails at the train function call and return the following traces:
```
`RuntimeError: Select TensorFlow op(s), included in the given model, is(are) not supported by this interpreter. 
Make sure you apply/link the Flex delegate before inference. For the Android, it can be resolved by adding "org.tensorflow:tensorflow-lite-select-tf-ops" dependency. 
See instructions: https://www.tensorflow.org/lite/guide/ops_select
Node number 54 (FlexReluGrad) failed to prepare.`
```
**Describe the expected behavior**
The expected behavior is to get a complete and functional training on target. 

Any help would be appreciated, thank you in advance.

**Comment by ahlzouao** [comment]
Another (probably helpful) and relevant information, delegate/flex sources seems to integrated in the  **`_pywrap_tensorflow_interpreter_wrapper.so`** as it can be shown below:
```
root@rpi:~# nm -a /usr/lib/python3.8/site-packages/tflite_runtime/_pywrap_tensorflow_interpreter_wrapper.so   | grep -i "Flex"
001552c8 W _ZN11flexbuffers10KeyCompareIhEEiPKvS2_
001552d8 W _ZN11flexbuffers10KeyCompareIjEEiPKvS2_
001552d0 W _ZN11flexbuffers10KeyCompareItEEiPKvS2_
001552c0 W _ZN11flexbuffers10KeyCompareIyEEiPKvS2_
0011a360 W _ZN6tflite19AcquireFlexDelegateEv
00123a74 T _ZN6tflite8IsFlexOpEPKc
002c3194 r _ZN6tfliteL21kFlexCustomCodePrefixE
001557b0 W _ZNK11flexbuffers9Reference7AsInt64Ev
001f4300 W _ZNK11flexbuffers9Reference8AsDoubleEv
002c4860 u _ZZN11flexbuffers3Map8EmptyMapEvE9empty_map
0011a1bc t _ZZN6tflite19AcquireFlexDelegateEvENUlP14TfLiteDelegateE_4_FUNES1_
```

**Comment by tilakrayal** [comment]
@ahlzouao ,
I was facing different error while executing the code.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/ebcb91fd4d6fd224a5cf843144e506aa/untitled210.ipynb).Please provide the complete code and dependencies to reproduce the issue.Thanks!

**Comment by ahlzouao** [comment]
@tilakrayal ,
Thank your for your answer, I have prepared a a gist [here](https://colab.research.google.com/drive/1rzvK_FtR0NZ0YUH9ya0L3uOoY3ZiZUM5?usp=sharing) with the whole code. 

Unfortunately, it is working just fine in the google colab environment and the train function returns the expected value, so that's not really the way to go. If you want to reproduce exactly the same issue, I would suggest you run the code on target by doing the following:
1. Run the notebook in the link I sent you, 
2. then from **Files** section download the **fashion_mnist_model.tflite** file. 
3. Send this model file to your board (Raspberry Pi for example) via scp or any other file transfer protocol.
4. From your board's terminal, run the following command.
`BOARD$ pip3 install tflite-runtime`
5. Then you will need to open a Python prompt 
`BOARD$ python3`
6. Type the following instructions:
```
import numpy as np
import struct as st
import tflite_runtime.interpreter as tflr
images_array = np.zeros((1, 28, 28), dtype="float32") 
labels_array = np.zeros((1,10), dtype="float32")
modelpath="fashion_mnist_model.tflite"
interpreter = tflr.Interpreter(modelpath)
signatures = interpreter.get_signature_list()
print('Signature:', signatures)
train = interpreter.get_signature_runner('train')
infer = interpreter.get_signature_runner('infer')
train(x=images_array, y=labels_array)
```
That's the output showing the error:
```
`RuntimeError: Select TensorFlow op(s), included in the given model, is(are) not supported by this interpreter. 
Make sure you apply/link the Flex delegate before inference. For the Android, it can be resolved by adding "org.tensorflow:tensorflow-lite-select-tf-ops" dependency. See instructions: https://www.tensorflow.org/lite/guide/ops_selectNode number 54 (FlexReluGrad) failed to prepare.`
```

**Comment by tilakrayal** [comment]
@ahlzouao ,
I was using colab to reproduce the issue.Can you please provide the  tf-lite file along with the code which you are facing issue.It helps to debug the issue.Thanks!

**Comment by ahlzouao** [comment]
@tilakrayal ,
The issue is not reproduceable on google colab, it will work as expected with no errors. You will need to run it on an armv7 or aarch64 architecture target to reproduce the issue, since the goal of this is to do on-device-training. [Here](https://drive.google.com/file/d/1i_emRlbfM_cSVltgkP81JqiN1rltM5Et/view?usp=sharing) is a link to the tf-lite file.

**Comment by tilakrayal** [comment]
@sachinprasadhs ,
I was able to execute the code in colab without any issues.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/d4261dcd1bc6f900a1ff720b5e0205c5/untitled213.ipynb).

**Comment by ahlzouao** [comment]
@tilakrayal @sachinprasadhs 
I already mentioned that the code will run without any issues on Google Colab and the issue is not reproduce-able on its environement, I'm actually facing the issue while running the code on device (on a raspberry Pi aarch64 architecture). The bug might be in the tflite-runtime python wheel of aarch64.

**Comment by ahlzouao** [comment]
Hello @sachinprasadhs ,
Have you been able to reproduce the issue on some device ? I can confirm you that it is reproduce-able on both armv8 and armv7 architectures.
Thank you in advance,

**Comment by ahlzouao** [comment]
Hello @MeghnaNatraj , @sachinprasadhs 
 I'm doubting this is isn't a bug but more of a missing feature. I have tried an interesting experimentation, I tried running the same script on my x86 host machine with no tensorflow package installed, only the tflite-runtime Python pip_package installed, this reproduces the same issue. After installing Tensorflow package, the error traces disappear. This made me think about some operation features missing in the interpreter I build using CMake.
I was trying to add debug some traces into the tensorflow source code to be able to figure it out. I have reached some conclusions:

- The traces I added were mainly near the intepreter function and allowed me to notice that the FlexReluGrad had no prepare function as you may notice below: 
```
[DEBUG][PrepareOpsStartingAt]:[1003]: Node index: 54
[DEBUG][OpPrepare][957]:Entering function ... 
[DEBUG][OpPrepare][959]:Op having a null prepare function 
[DEBUG][OpPrepare][962]:Unresolved custom op ... 
```
- Based on the documentation, in order to do on-device-training, it is necessary to add the **Flex Delegate** to the interpreter during the build and select some tensorflow ops so the FlexOps could be understood by the interpreter (Hence the error, I'm getting each time : **_Node number 54 (FlexReluGrad) failed to prepare._** 
- By getting a closer look inside the libtensorflow-lite.a generated by  CMake, I can see that none of the flex delegate function has been integrated in the build since the variable [TFLITE_DELEGATES_FLEX_SRCS](https://github.com/tensorflow/tensorflow/blob/r2.7/tensorflow/lite/CMakeLists.txt#L399) is empty.
- By digging deeper inside the Bazel BUILD scripts, I was trying to duplicate the same [build dependencies](https://github.com/tensorflow/tensorflow/blob/r2.7/tensorflow/lite/delegates/flex/BUILD#L126) and sources to build using CMake. I have also noticed that many dependencies from tensorflow/core/framework are used.

=> To make it short, I'm trying to build a shared library containing the flex delegates with the _with_select_tf_ops=true_ but using **Cmake** instead of **Bazel** since it's not an option for me. This may be pushed as a new feature in the [tensorflow/lite/CMakeList.txt](https://github.com/tensorflow/tensorflow/blob/r2.7/tensorflow/lite/CMakeLists.txt)  upon completion, so I'm looking for some guidance on the dependencies and sources to build for that purpose.

**Comment by MeghnaNatraj** [comment]
@haozha111 triaging for on-device related issues.

**Comment by haozha111** [comment]
Hi,

Sorry for the late reply. ReluGrad isn't a supported built-in op of TF Lite. So you will need to link in the Flex delegate to make it able to run on-device.

**Comment by ahlzouao** [comment]
Hello @haozha111, 
First of all, thank you for your reply.
That's what I figured out, then I tried to build a new libtensorflow-lite-flex.so library with the Flex Delegates and the necessary TF Ops kernels (ReluOp, SaveOp and RestoreOp in this case) using SELECTIVE_REGISTRATION Mode, but the build does not seem to be working properly for aarch64 as stated in this [issue](https://github.com/tensorflow/tensorflow/issues/54500).

Also, do you have a clearer example on how to link the flex delegate to my interpreter ? I tried doing this : 
```
//Building the FlexDelegate instance
       auto delegate = tflite::FlexDelegate::Create();

       //Building the interpreter
       std::unique_ptr<tflite::Interpreter> interpreter;
       builder(&interpreter);

       // Modifying the graph with the FlexDelegate instance
       if(delegate){
            interpreter->ModifyGraphWithDelegate(std::move(delegate));
      }
```
For now, I cannot tell if it is working properly or not because I still don't have a properly working shared library libtensorflow-lite-flex.so to link against.

Thank you for your support,
Othmane

**Comment by haozha111** [comment]
Will https://github.com/tensorflow/tensorflow/issues/52018 help you resolve the build issue?

**Comment by ahlzouao** [comment]
Hello @haozha111, 
Sorry, I was mistaken in the link of the issue. The one I mentioned was actually closed and the #52018 did help, hence it was closed. The issue I'm facing during the build is related to some dependency "icu". A ticket has been opened in [here](https://github.com/tensorflow/tensorflow/issues/54517) but no answers so far. 

Thank you again for your support,
